### PR TITLE
Neutrinod layers

### LINKED ISSUE #18: README: Websocket/API usage documentation
Document how users should use instance's WebSocket interface

### LINKED ISSUE #19: Makefile: update make build command and add make run for development

### LINKED ISSUE #20: config/viper: add ENV variables to tweak listening address and elements host

### LINKED ISSUE #22: Split `neutrinod` in layers (core,app,ports, interface, infrastructure)

### LINKED ISSUE #23: Webhook: notify events to a registereded HTTP URL
When user subscribe and load his address/descriptor, we should allow user to be notified, beside via websocket, also to specify a webhook endpoint to be notified of new events

### PR DESCRIPTION
**Pull request by sekulicd**
This refactors neutrinod so that it aligned with layered arch.
NeutrinoD now supports multiple clients.
Node Start method refactored so that it blocks until full sync.

This closes #22 
This closes #18
This closes #19
This closes #20
This closes #23 

@tiero please review.

### CODE REVIEW

File: cmd/start.go [hunk 1]
```diff
@@ -73,7 +73,7 @@ func startAction(state *State) cli.ActionFunc {
 		}()
 
 		// we'll watch if this address receives fund
-		watchItem, err := scanner.NewScriptWatchItemFromAddress(addr)
+		watchItem, err := scanner.NewUnspentWatchItemFromAddress(addr)
```
Comment (tiero) [inline cmd/start.go]: seems to me cmd/start.go not needed anymore? since we use `neutrinod`

Comment (sekulicd) [reply cmd/start.go]: Done

File: README.md [hunk 2]
```diff
@@ -10,30 +10,49 @@ Neutrino + Elements
...
+- `ScannerService` uses filters and headers repositories to handle `ScanRequest` which aims to know if an outpoint (identified by its script) is spent or not.<br>
 
-## Getting Started
+Two binaries are provided if you want to use ready light client:<br>
+- `neutrinod` is a daemon that accepts websocket connections on which clients can send requests to watch for events related to wallet-descriptor<br>
+neutrinod also exposes HTTP endpoint that supports adding webhooks to receive notifications when an event occurs.<br>
+- `neutrino` is a simple command line tool that can be used to watch Liquid side-chain events.<br>
 
-### Build
+## Usage
```
Comment (tiero) [inline README.md]: Divide development from usage of user of pre-built binary

File: README.md [hunk 3]
```diff
@@ -10,30 +10,49 @@ Neutrino + Elements
...
+### Start neutrinod
 
 ```
-make test
+./bin/neutrinod
 ```
 
-### Format
+### Use neutrino CLI
+
```
Comment (tiero) [inline README.md:47]: Add instruction to download from Github Releases

File: README.md [hunk 4]
```diff
@@ -10,30 +10,49 @@ Neutrino + Elements
...
+#### Config CLI
+```
+./bin/neutrino config
+```
 
+#### Watch for events related to wallet-descriptor(websocket-{HOST:PORT}/neutrino/subscribe/ws)
```
Comment (tiero) [inline README.md]: split the HTTP API doc in a dedicated paragraph

File: README.md [hunk 5]
```diff
@@ -22,37 +22,81 @@ Two binaries are provided if you want to use ready light client:<br>
...
+make test
 ```
 
-### Use neutrino CLI
+## Usage
+
+### Run neutrinod
+
+```
+./bin/neutrinod
```
Comment (tiero) [inline README.md]: end user will likely use the CLI as binary already built, ie. downloaded from Github Releases

**Comment by tiero** [comment]
@sekulicd 
I am running it with the following descriptor, generated with Marina, with nigiri locally. The connection closes 

```bash
./bin/neutrino subscribe --descriptor="wpkh(xpub661MyMwAqRbcGjSvn4mmnGq6joEwo4BA5beEbRytbcNfCEcJqUqZJRpsbn9MSbRtmQKhcVMJaeL6sDsV7QnuRRgNogaTRTSqLjzQyQHg9yD)" --block_height=0
INFO[0000] successfully registered
[neutrino] websocket: close 1006 (abnormal closure): unexpected EOF
```

on the logs I get

```bash
INFO[1558] sucesfull registration, subscriber: 57cf386c-49be-4a3c-95d3-8f6d5c0edf66
WARN[1618] Error reading message: read tcp 127.0.0.1:8080->127.0.0.1:61769: i/o timeout
INFO[1618] subscriber: 57cf386c-49be-4a3c-95d3-8f6d5c0edf66, un-registrated
ERRO[1618] failed unsubscribing from chain: subscriber not found
```


PS: I replicated same problem with Postman WebSocket client

File: internal/interface/web-socket/handler/notification_handler.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+func (d *descriptorWalletNotifierHandler) handleSubscribers() {
+	for {
+		select {
+		case sub := <-d.registerSubs:
+			log.Debugf("subscriber: %v, registrated", uuid.UUID(sub.SubscriberID()).String())
+			d.addSubscriberSafe(sub)
+
+		case sub := <-d.unregisterSubs:
+			log.Infof("subscriber: %v, un-registrated", uuid.UUID(sub.SubscriberID()).String())
```
Comment (tiero) [inline internal/interface/web-socket/handler/notification_handler.go]:
```suggestion
			log.Infof("subscriber: %v, un-registered", uuid.UUID(sub.SubscriberID()).String())
```

File: README.md [hunk 7]
```diff
@@ -10,30 +10,93 @@ Neutrino + Elements
...
+curl -X POST http://localhost:8080/neutrino/subscribe/http \
+   -H 'Content-Type: application/json' \
+   -d '{"actionType":"register", "eventTypes":[0], "descriptorWallet":"wpkh(037470e26cc774eca62ca19e1a182461a5f3d3680acbc593ce3f38cd142c26c03d)", "startBlockHeight":0, "endpointUrl":"http://127.0.0.1:62900"}'
+```
+
+## API
+Neutrinod can be used to subscribe to events related to wallet-descriptor using web-socket or by registrating webhook using HTTP.<br>
 
+### Web-socket
+To subscribe to events related to wallet-descriptor using web-socket connection, send bellow json :<br>
```
Comment (tiero) [inline README.md]:
```suggestion
To subscribe to events related to wallet-descriptor using web-socket connection, send below json :<br>
```

File: README.md [hunk 8]
```diff
@@ -10,30 +10,93 @@ Neutrino + Elements
...
+```
```
Comment (tiero) [inline README.md]:
```suggestion
```json
```

File: README.md [hunk 9]
```diff
@@ -10,30 +10,93 @@ Neutrino + Elements
...
+}
+```
+to the following endpoint:
+```
+{HOST}:8080/neutrino/subscribe/ws
+```
+
+### HTTP 
+To register webhook, send bellow json :<br>
+```
```
Comment (tiero) [inline README.md]:
```suggestion
```json
```

File: README.md [hunk 10]
```diff
@@ -10,30 +10,93 @@ Neutrino + Elements
...
+
+### Watch for events related to wallet-descriptor(websocket-{HOST:PORT}/neutrino/subscribe/ws) 
+```
+./bin/neutrino subscribe --descriptor="{WALLET_DESCRIPTOR}" --block_height={BLOCK_HEIGHT}
+```
+### Add webhook(http)
+```
+curl -X POST http://localhost:8080/neutrino/subscribe/http \
+   -H 'Content-Type: application/json' \
+   -d '{"actionType":"register", "eventTypes":[0], "descriptorWallet":"wpkh(037470e26cc774eca62ca19e1a182461a5f3d3680acbc593ce3f38cd142c26c03d)", "startBlockHeight":0, "endpointUrl":"http://127.0.0.1:62900"}'
```
Comment (tiero) [inline README.md]: eventTypes passed as number are not documented anywhere. I think is best to have human readable string here in the user-facing API

File: README.md [hunk 11]
```diff
@@ -10,30 +10,93 @@ Neutrino + Elements
...
+```
+./bin/neutrino config
+```
+
+### Watch for events related to wallet-descriptor(websocket-{HOST:PORT}/neutrino/subscribe/ws) 
+```
+./bin/neutrino subscribe --descriptor="{WALLET_DESCRIPTOR}" --block_height={BLOCK_HEIGHT}
+```
+### Add webhook(http)
+```
```
Comment (tiero) [inline README.md]:
```suggestion
```sh
```

File: internal/config/config.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+var (
+	vip *viper.Viper
+)
+
+func LoadConfig() error {
+	vip = viper.New()
+	vip.SetEnvPrefix("NEUTRINO")
+	vip.AutomaticEnv()
+
+	vip.SetDefault(NeutrinoDUrlKey, "localhost:8080")
```
Comment (tiero) [inline internal/config/config.go]: 8080 is usually used by frontends

```suggestion
	vip.SetDefault(NeutrinoDUrlKey, "localhost:8000")
```

File: README.md [hunk 13]
```diff
@@ -10,30 +10,93 @@ Neutrino + Elements
...
+    "actionType":"register", 
+    "eventTypes":[0], 
+    "descriptorWallet":"{WALLET_DESCRIPTOR}", 
+    "startBlockHeight":{START_BLOCK_HEIGHT}, 
+    "endpointUrl":"{ENDPOINT_URL}"
+}
+```
+to the following endpoint:
+```
+{HOST}:8080/neutrino/subscribe/ws
```
Comment (tiero) [inline README.md]: Being user facing, you can omit the endpoint, that would depend on deployment and we do not have control on the actual port

```suggestion
/neutrino/subscribe/ws
```

File: README.md [hunk 14]
```diff
@@ -10,30 +10,93 @@ Neutrino + Elements
...
 
+### Web-socket
+To subscribe to events related to wallet-descriptor using web-socket connection, send bellow json :<br>
+```
+{
+    "actionType":"register", 
+    "eventTypes":[0], 
+    "descriptorWallet":"{WALLET_DESCRIPTOR}", 
+    "startBlockHeight":{START_BLOCK_HEIGHT}, 
+    "endpointUrl":"{ENDPOINT_URL}"
```
Comment (tiero) [inline README.md]: What this `endpointUrl` is used for in websocket?

Comment (sekulicd) [reply README.md]: removed

File: README.md (hunk 13 above)
Comment (sekulicd) [reply README.md]: Done

File: internal/config/config.go (hunk 12 above)
Comment (sekulicd) [reply internal/config/config.go]: Done

File: README.md (hunk 11 above)
Comment (sekulicd) [reply README.md]: Done

File: README.md (hunk 10 above)
Comment (sekulicd) [reply README.md]: Done

File: README.md (hunk 9 above)
Comment (sekulicd) [reply README.md]: Done

File: README.md (hunk 8 above)
Comment (sekulicd) [reply README.md]: Done

File: README.md (hunk 7 above)
Comment (sekulicd) [reply README.md]: Done

File: internal/interface/web-socket/handler/notification_handler.go (hunk 6 above)
Comment (sekulicd) [reply internal/interface/web-socket/handler/notification_handler.go]: Done

**Comment by sekulicd** [comment]
> @sekulicd I am running it with the following descriptor, generated with Marina, with nigiri locally. The connection closes
> 
> ```shell
> ./bin/neutrino subscribe --descriptor="wpkh(xpub661MyMwAqRbcGjSvn4mmnGq6joEwo4BA5beEbRytbcNfCEcJqUqZJRpsbn9MSbRtmQKhcVMJaeL6sDsV7QnuRRgNogaTRTSqLjzQyQHg9yD)" --block_height=0
> INFO[0000] successfully registered
> [neutrino] websocket: close 1006 (abnormal closure): unexpected EOF
> ```
> 
> on the logs I get
> 
> ```shell
> INFO[1558] sucesfull registration, subscriber: 57cf386c-49be-4a3c-95d3-8f6d5c0edf66
> WARN[1618] Error reading message: read tcp 127.0.0.1:8080->127.0.0.1:61769: i/o timeout
> INFO[1618] subscriber: 57cf386c-49be-4a3c-95d3-8f6d5c0edf66, un-registrated
> ERRO[1618] failed unsubscribing from chain: subscriber not found
> ```
> 
> PS: I replicated same problem with Postman WebSocket client

Connection timeout removed

**Pull request closed without merging**